sklearn/kernel_approximation.py

```diff
@@ -22,7 +24,158 @@
+
+    [2] Charikar, M., Chen, K., & Farach-Colton, M. (2002). Finding frequent
+    items in data streams. In International Colloquium on Automata,
+    Languages, and Programming (pp. 693-703). Springer, Berlin, Heidelberg.
+    (http://www.vldb.org/pvldb/1/1454225.pdf)
+
+    """
+
+    def __init__(self, degree=2, coef0=0, n_components=100, random_state=None):
+
+        self.degree = degree
+        self.coef0 = coef0
+        self.n_components = n_components
+        self.random_state = random_state
+
+    def fit(self, X, y=None):
+        """Fit the model with X.
+
+        Initializes the internal variables. The method is data-independent,
+        so we only care about the dimension of samples in X.
+
+        Parameters
+        ----------
+        X : {array-like, sparse matrix} of shape (n_samples, n_features)
+            Training data, where n_samples in the number of samples
+            and n_features is the number of features.
+
+        Returns
+        -------
+        self : object
+            Returns the transformer.
+        """
+
+        X = check_array(X, accept_sparse="csc")
+        random_state = check_random_state(self.random_state)
+
+        if sp.issparse(X) and self.coef0 != 0:
+            X = sp.hstack([X, np.sqrt(self.coef0) * np.ones((X.shape[0], 1))],
+                          format="csc")
+
+        elif not sp.issparse(X) and self.coef0 != 0:
+            X = np.hstack([X, np.sqrt(self.coef0)*np.ones((X.shape[0], 1))])
+
+        n_features = X.shape[1]
+
+        self.indexHash_ = random_state.randint(0, high=self.n_components,
+                                               size=(self.degree, n_features))
+
+        self.bitHash_ = random_state.choice(a=[-1., 1.],
+                                            size=(self.degree, n_features)
+                                            ).astype(np.float32)
+        return self
+
+    def transform(self, X):
+        """Generate the feature map approximation for X.
+
+        Parameters
+        ----------
+        X : {array-like}, shape (n_samples, n_features)
+            New data, where n_samples in the number of samples
+            and n_features is the number of features.
+
+        Returns
+        -------
+        X_new : array-like, shape (n_samples, n_components)
+        """
+
+        check_is_fitted(self)
+        X = check_array(X, accept_sparse="csc")
+
+        if sp.issparse(X) and self.coef0 != 0:
+            X = sp.hstack([X, np.sqrt(self.coef0)*np.ones((X.shape[0], 1))],
+                          format="csc")
+
+        elif not sp.issparse(X) and self.coef0 != 0:
+            X = np.hstack([X, np.sqrt(self.coef0)*np.ones((X.shape[0], 1))])
+
+        if X.shape[1] != self.indexHash_.shape[1]:
+            raise ValueError("Number of features of test samples does not"
+                             " match that of training samples.")
+
+        Ps = np.zeros((X.shape[0], self.degree, self.n_components))
```

Comment:
What is Ps ? We try to avoid variable names without meaning. If you think it is still better this way, you could add a comment to state what it means.

Reply:
- [x] Give intermediate variables more meaningful names.